File: frontend/types/discount.ts
// Shape response backend modul discount (CLAUDE.md §28). Sinkronkan kalau
// backend berubah.

export type DiscountType = 'percent' | 'nominal' | 'nominal_per_m2'
export type DiscountChannelScope = 'all' | 'online' | 'pos'
export type DiscountStatus = 'aktif' | 'terjadwal' | 'kadaluarsa' | 'nonaktif' | 'kuota_habis'
/** Cakupan produk (CLAUDE.md §28.9). `selected` TANPA `product_ids` bukan "berlaku semua" — itu diskon yang tidak bisa dipakai sama sekali. */
export type DiscountAppliesTo = 'all' | 'selected'
/** Cakupan audiens (CLAUDE.md §30.3). `member` menambah syarat customer harus berstatus `active` (lihat useMembership). */
export type DiscountAudienceScope = 'all' | 'member'
/**
 * Hanya relevan kalau `audience_scope === 'member'`. BEDA dengan `applies_to`
 * (§28.9): ini enum eksplisit, bukan "kosong = semua" — `selected_members`
 * dengan `customer_ids` kosong WAJIB ditolak backend (§30.3), jangan
 * disamakan dengan pola `applies_to==='selected'`.
 */
export type DiscountMemberScope = 'all_members' | 'selected_members' | ''

export interface Discount {
  id: string
  code: string
  name: string
  type: DiscountType
  value_percent: number | null
  value_amount: number | null
  max_discount_amount: number | null
  min_subtotal: number
  starts_at: string | null
  ends_at: string | null
  quota: number | null
  usage_count: number
  channel_scope: DiscountChannelScope
  is_active: boolean
  status: DiscountStatus
  /** Default `all` — sinkron dengan default backend (§28.9). */
  applies_to: DiscountAppliesTo
  /** Kosong kalau `applies_to === 'all'`. */
  product_ids: string[]
  /** Default `all` — sinkron dengan default backend (§30.3). */
  audience_scope: DiscountAudienceScope
  /** "" kalau `audience_scope === 'all'`. */
  member_scope: DiscountMemberScope
  /** Kosong kalau `member_scope !== 'selected_members'`. */
  customer_ids: string[]
  created_at: string
  updated_at: string
}

/** Hasil `GET /admin/discounts/applicable` — Discount + estimasi potongan untuk subtotal yang dikirim. */
export interface ApplicableDiscount extends Discount {
  preview_amount: number
}

export interface DiscountListResponse {
  items: Discount[]
  total: number
  page: number
  per_page: number
}

/**
 * Body create/update. Field kosong/undefined dianggap "tanpa batas" oleh
 * backend UNTUK `max_discount_amount`/`starts_at`/`ends_at`/`quota` — kirim
 * `null` eksplisit untuk mengosongkannya lewat PATCH.
 *
 * `value_percent`/`value_amount` BEDA ATURAN: PATCH backend (`parseUpdateInput`)
 * menolak keras `null` eksplisit untuk dua field ini (cuma menerima "field
 * absen" atau "field berisi angka") — caller WAJIB mengirim `undefined`
 * (bukan `null`) kalau field itu tidak relevan untuk `type` yang dipilih.
 */
export interface DiscountInput {
  code: string
  name: string
  type: DiscountType
  value_percent?: number | null
  value_amount?: number | null
  max_discount_amount?: number | null
  min_subtotal: number
  starts_at?: string | null
  ends_at?: string | null
  quota?: number | null
  channel_scope: DiscountChannelScope
  is_active: boolean
  applies_to: DiscountAppliesTo
  /** Wajib diisi (minimal 1) kalau `applies_to === 'selected'` — backend menolak daftar kosong (§28.9). */
  product_ids: string[]
  audience_scope: DiscountAudienceScope
  /** Wajib diisi ('all_members'/'selected_members') kalau `audience_scope === 'member'`, wajib "" kalau tidak (§30.3). */
  member_scope: DiscountMemberScope
  /** Wajib diisi (minimal 1) kalau `member_scope === 'selected_members'` — backend menolak daftar kosong (§30.3). */
  customer_ids: string[]
}

File: backend/internal/discount/model/discount.go
// Package model contains GORM entities for the discount module (§28).
package model

import (
	"time"

	"github.com/google/uuid"
)

// DiscountType — tipe kalkulasi diskon (§28.1).
type DiscountType string

const (
	DiscountTypePercent      DiscountType = "percent"
	DiscountTypeNominal      DiscountType = "nominal"
	DiscountTypeNominalPerM2 DiscountType = "nominal_per_m2"
	// DiscountTypeManual — dipakai HANYA di kolom snapshot orders
	// (discount_type_snapshot), tidak pernah tersimpan sebagai baris
	// discounts.type (diskon manual tidak punya master row, discount_id nil).
	DiscountTypeManual DiscountType = "manual"
)

// ChannelScope — batas channel order tempat diskon boleh dipakai (§28.1).
type ChannelScope string

const (
	ChannelScopeAll    ChannelScope = "all"
	ChannelScopeOnline ChannelScope = "online"
	ChannelScopePOS    ChannelScope = "pos"
)

// AppliesToScope — cakupan produk sebuah diskon (§28.9). Default 'all' wajib
// membuat diskon lama (dibuat sebelum kolom ini ada) tetap berperilaku
// persis seperti sebelumnya. 'selected' berarti diskon HANYA boleh dipakai
// untuk produk yang terdaftar di tabel discount_products — dan daftar
// kosong TIDAK PERNAH berarti "berlaku untuk semua" (lihat discountapi
// ErrDiscountScopeEmpty).
type AppliesToScope string

const (
	AppliesToAll      AppliesToScope = "all"
	AppliesToSelected AppliesToScope = "selected"
)

// AudienceScope — batas audiens sebuah diskon (§30.3): apakah diskon berlaku
// untuk semua orang, atau khusus member. Independen dari AppliesToScope
// (cakupan produk) — sebuah diskon boleh sekaligus "khusus member" DAN
// "khusus produk tertentu". Default 'all' wajib membuat diskon lama (dibuat
// sebelum kolom ini ada) tetap berperilaku persis seperti sebelumnya.
type AudienceScope string

const (
	AudienceScopeAll    AudienceScope = "all"
	AudienceScopeMember AudienceScope = "member"
)

// MemberScope — hanya relevan kalau AudienceScope == AudienceScopeMember.
// BEDA dengan AppliesToScope: ini enum eksplisit, bukan disimpulkan dari isi
// discount_customers kosong-atau-tidak (§30.3) — 'all_members' berarti
// berlaku untuk SEMUA member aktif (discount_customers tidak dipakai sama
// sekali, boleh kosong, itu normal); 'selected_members' berarti berlaku
// HANYA untuk customer yang terdaftar di discount_customers, dan daftar
// kosong TIDAK PERNAH berarti "berlaku untuk semua" (lihat discountapi
// ErrDiscountMemberScopeEmpty).
type MemberScope string

const (
	MemberScopeAllMembers MemberScope = "all_members"
	MemberScopeSelected   MemberScope = "selected_members"
)

// Discount — satu baris = satu program diskon (§28.1). TIDAK PERNAH
// di-hard-delete — soft delete via DeletedAt/DeletedBy/DeleteReason, pola
// sama seperti order.model.Order (migration 000025).
type Discount struct {
	ID                uuid.UUID      `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	Code              string         `gorm:"size:30;not null;column:code"                    json:"code"`
	Name              string         `gorm:"size:150;not null;column:name"                   json:"name"`
	Type              DiscountType   `gorm:"size:20;not null;column:type"                    json:"type"`
	ValuePercent      *float64       `gorm:"column:value_percent"                            json:"value_percent"`
	ValueAmount       *int64         `gorm:"column:value_amount"                             json:"value_amount"`
	MaxDiscountAmount *int64         `gorm:"column:max_discount_amount"                      json:"max_discount_amount"`
	MinSubtotal       int64          `gorm:"not null;default:0;column:min_subtotal"          json:"min_subtotal"`
	StartsAt          *time.Time     `gorm:"column:starts_at"                                json:"starts_at"`
	EndsAt            *time.Time     `gorm:"column:ends_at"                                  json:"ends_at"`
	Quota             *int           `gorm:"column:quota"                                    json:"quota"`
	ChannelScope      ChannelScope   `gorm:"size:20;not null;default:all;column:channel_scope" json:"channel_scope"`
	AppliesTo         AppliesToScope `gorm:"size:20;not null;default:all;column:applies_to"  json:"applies_to"`
	AudienceScope     AudienceScope  `gorm:"size:20;not null;default:all;column:audience_scope" json:"audience_scope"`
	MemberScope       *MemberScope   `gorm:"size:20;column:member_scope"                     json:"member_scope"`
	IsActive          bool           `gorm:"not null;default:true;column:is_active"          json:"is_active"`

	DeletedAt    *time.Time `gorm:"column:deleted_at"    json:"deleted_at,omitempty"`
	DeletedBy    *uuid.UUID `gorm:"type:uuid;column:deleted_by" json:"deleted_by,omitempty"`
	DeleteReason *string    `gorm:"column:delete_reason" json:"delete_reason,omitempty"`

	CreatedAt time.Time `gorm:"not null;default:now();column:created_at" json:"created_at"`
	UpdatedAt time.Time `gorm:"not null;default:now();column:updated_at" json:"updated_at"`
}

func (Discount) TableName() string { return "discounts" }
